File: dqops/src/main/java/com/dqops/connectors/snowflake/SnowflakeParametersSpec.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.connectors.snowflake;

import com.dqops.connectors.ConnectionProviderSpecificParameters;
import com.dqops.core.secrets.SecretValueLookupContext;
import com.dqops.core.secrets.SecretValueProvider;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.sources.BaseProviderParametersSpec;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;
import picocli.CommandLine;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;

/**
 * Snowflake connection parameters.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class SnowflakeParametersSpec extends BaseProviderParametersSpec
        implements ConnectionProviderSpecificParameters {
    private static final ChildHierarchyNodeFieldMapImpl<SnowflakeParametersSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(BaseProviderParametersSpec.FIELDS) {
        {
        }
    };

    @CommandLine.Option(names = {"--snowflake-account"}, description = "Snowflake account name, e.q. <account>, <account>-<locator>, <account>.<region> or <account>.<region>.<platform>.")
    @JsonPropertyDescription("Snowflake account name, e.q. <account>, <account>-<locator>, <account>.<region> or <account>.<region>.<platform>.. Supports also a ${SNOWFLAKE_ACCOUNT} configuration with a custom environment variable.")
    private String account;

    @CommandLine.Option(names = {"--snowflake-warehouse"}, description = "Snowflake warehouse name.")
    @JsonPropertyDescription("Snowflake warehouse name. Supports also a ${SNOWFLAKE_WAREHOUSE} configuration with a custom environment variable.")
    private String warehouse;

    @CommandLine.Option(names = {"--snowflake-database"}, description = "Snowflake database name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    @JsonPropertyDescription("Snowflake database name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    private String database;

    @CommandLine.Option(names = {"--snowflake-user"}, description = "Snowflake user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    @JsonPropertyDescription("Snowflake user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    private String user;

    @CommandLine.Option(names = {"--snowflake-password"}, description = "Snowflake database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    @JsonPropertyDescription("Snowflake database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    private String password;

    @CommandLine.Option(names = {"--snowflake-role"}, description = "Snowflake role name.")
    @JsonPropertyDescription("Snowflake role name. Supports also ${SNOWFLAKE_ROLE} configuration with a custom environment variable.")
    private String role;

    @CommandLine.Option(names = {"-F"}, description = "Snowflake additional properties that are added to the JDBC connection string")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String, String> properties;

    /**
     * Returns a snowflake account name.
     * @return Snowflake account name.
     */
    public String getAccount() {
        return account;
    }

    /**
     * Sets a snowflake account name.
     * @param account Snowflake account name.
     */
    public void setAccount(String account) {
		setDirtyIf(!Objects.equals(this.account, account));
        this.account = account;
    }

    /**
     * Returns the snowflake warehouse name.
     * @return Warehouse name.
     */
    public String getWarehouse() {
        return warehouse;
    }

    /**
     * Sets the snowflake warehouse name.
     * @param warehouse Warehouse name.
     */
    public void setWarehouse(String warehouse) {
		setDirtyIf(!Objects.equals(this.warehouse, warehouse));
        this.warehouse = warehouse;
    }

    /**
     * Returns a physical database name.
     * @return Physical database name.
     */
    public String getDatabase() {
        return database;
    }

    /**
     * Sets a physical database name.
     * @param database Physical database name.
     */
    public void setDatabase(String database) {
        setDirtyIf(!Objects.equals(this.database, database));
        this.database = database;
    }

    /**
     * Returns the user that is used to log in to the data source (JDBC user or similar).
     * @return User name.
     */
    public String getUser() {
        return user;
    }

    /**
     * Sets a user name.
     * @param user User name.
     */
    public void setUser(String user) {
        setDirtyIf(!Objects.equals(this.user, user));
        this.user = user;
    }

    /**
     * Returns a password used to authenticate to the server.
     * @return Password.
     */
    public String getPassword() {
        return password;
    }

    /**
     * Sets a password that is used to connect to the database.
     * @param password Password.
     */
    public void setPassword(String password) {
        setDirtyIf(!Objects.equals(this.password, password));
        this.password = password;
    }

    /**
     * Gets the role.
     * @return Role name.
     */
    public String getRole() {
        return role;
    }

    /**
     * Sets the role name.
     * @param role Role name.
     */
    public void setRole(String role) {
		setDirtyIf(!Objects.equals(this.role, role));
        this.role = role;
    }

    /**
     * Returns a key/value map of additional properties that are included in the JDBC connection string.
     * @return Key/value dictionary of additional JDBC properties.
     */
    public Map<String, String> getProperties() {
        return properties;
    }

    /**
     * Sets a dictionary of additional connection parameters that are added to the JDBC connection string.
     * @param properties Key/value dictionary with extra parameters.
     */
    public void setProperties(Map<String, String> properties) {
        setDirtyIf(!Objects.equals(this.properties, properties));
        this.properties = properties != null ? Collections.unmodifiableMap(properties) : null;
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Creates and returns a deep copy of this object.
     */
    @Override
    public SnowflakeParametersSpec deepClone() {
        SnowflakeParametersSpec cloned = (SnowflakeParametersSpec)super.deepClone();
        return cloned;
    }

    /**
     * Creates a trimmed and expanded version of the object without unwanted properties, but with all variables like ${ENV_VAR} expanded.
     * @param secretValueProvider Secret value provider.
     * @param lookupContext Secret lookup context.
     * @return Trimmed and expanded version of this object.
     */
    public SnowflakeParametersSpec expandAndTrim(SecretValueProvider secretValueProvider, SecretValueLookupContext lookupContext) {
        SnowflakeParametersSpec cloned = this.deepClone();
        cloned.account = secretValueProvider.expandValue(cloned.account, lookupContext);
        cloned.warehouse = secretValueProvider.expandValue(cloned.warehouse, lookupContext);
        cloned.database = secretValueProvider.expandValue(cloned.database, lookupContext);
        cloned.user = secretValueProvider.expandValue(cloned.user, lookupContext);
        cloned.password = secretValueProvider.expandValue(cloned.password, lookupContext);
        cloned.role = secretValueProvider.expandValue(cloned.role, lookupContext);
        cloned.properties = secretValueProvider.expandProperties(cloned.properties, lookupContext);

        return cloned;
    }
}

File: distribution/python/dqops/client/models/snowflake_parameters_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.snowflake_parameters_spec_properties import (
        SnowflakeParametersSpecProperties,
    )


T = TypeVar("T", bound="SnowflakeParametersSpec")


@_attrs_define
class SnowflakeParametersSpec:
    """
    Attributes:
        account (Union[Unset, str]): Snowflake account name, e.q. <account>, <account>-<locator>, <account>.<region> or
            <account>.<region>.<platform>.. Supports also a ${SNOWFLAKE_ACCOUNT} configuration with a custom environment
            variable.
        warehouse (Union[Unset, str]): Snowflake warehouse name. Supports also a ${SNOWFLAKE_WAREHOUSE} configuration
            with a custom environment variable.
        database (Union[Unset, str]): Snowflake database name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME}
            format to use dynamic substitution.
        user (Union[Unset, str]): Snowflake user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to
            use dynamic substitution.
        password (Union[Unset, str]): Snowflake database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME}
            format to use dynamic substitution.
        role (Union[Unset, str]): Snowflake role name. Supports also ${SNOWFLAKE_ROLE} configuration with a custom
            environment variable.
        properties (Union[Unset, SnowflakeParametersSpecProperties]): A dictionary of custom JDBC parameters that are
            added to the JDBC connection string, a key/value dictionary.
    """

    account: Union[Unset, str] = UNSET
    warehouse: Union[Unset, str] = UNSET
    database: Union[Unset, str] = UNSET
    user: Union[Unset, str] = UNSET
    password: Union[Unset, str] = UNSET
    role: Union[Unset, str] = UNSET
    properties: Union[Unset, "SnowflakeParametersSpecProperties"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        account = self.account
        warehouse = self.warehouse
        database = self.database
        user = self.user
        password = self.password
        role = self.role
        properties: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.properties, Unset):
            properties = self.properties.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if account is not UNSET:
            field_dict["account"] = account
        if warehouse is not UNSET:
            field_dict["warehouse"] = warehouse
        if database is not UNSET:
            field_dict["database"] = database
        if user is not UNSET:
            field_dict["user"] = user
        if password is not UNSET:
            field_dict["password"] = password
        if role is not UNSET:
            field_dict["role"] = role
        if properties is not UNSET:
            field_dict["properties"] = properties

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.snowflake_parameters_spec_properties import (
            SnowflakeParametersSpecProperties,
        )

        d = src_dict.copy()
        account = d.pop("account", UNSET)

        warehouse = d.pop("warehouse", UNSET)

        database = d.pop("database", UNSET)

        user = d.pop("user", UNSET)

        password = d.pop("password", UNSET)

        role = d.pop("role", UNSET)

        _properties = d.pop("properties", UNSET)
        properties: Union[Unset, SnowflakeParametersSpecProperties]
        if isinstance(_properties, Unset):
            properties = UNSET
        else:
            properties = SnowflakeParametersSpecProperties.from_dict(_properties)

        snowflake_parameters_spec = cls(
            account=account,
            warehouse=warehouse,
            database=database,
            user=user,
            password=password,
            role=role,
            properties=properties,
        )

        snowflake_parameters_spec.additional_properties = d
        return snowflake_parameters_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
